src/concurrency/ocf_pio_concurrency.c

```diff
@@ -0,0 +1,228 @@
+		if (entry == OUT_OF_RANGE)
+			continue;
+
+		ENV_BUG_ON(ocf_alock_is_index_locked(alock, req, i));
+
+		if (ocf_alock_trylock_entry_wr(alock, entry)) {
+			ocf_alock_mark_index_locked(alock, req, i, true);
+		} else {
+			ret = OCF_LOCK_NOT_ACQUIRED;
+			break;
+		}
+	}
+
+	if (ret != OCF_LOCK_NOT_ACQUIRED)
+		return ret;
+
+	/* Request is not locked, discard acquired locks */
+	for (; i >= 0; i--) {
+		entry = ocf_pio_lock_get_entry(alock, req, i);
+		if (entry == OUT_OF_RANGE)
+			continue;
+
+		if (ocf_alock_is_index_locked(alock, req, i)) {
+			ocf_alock_unlock_one_wr(alock, entry);
+			ocf_alock_mark_index_locked(alock, req, i, false);
+		}
+	}
+
+	return ret;
+}
+
+static int ocf_pio_lock_slow(struct ocf_alock *alock,
+		struct ocf_request *req, int rw, ocf_req_async_lock_cb cmpl)
+{
+	int32_t i;
+	ocf_cache_line_t entry;
+	int ret = OCF_LOCK_ACQUIRED;
+	ENV_BUG_ON(rw != OCF_WRITE);
+
+	for (i = 0; i < req->core_line_count; i++) {
+		entry = ocf_pio_lock_get_entry(alock, req, i);
+		if (entry == OUT_OF_RANGE)
+			continue;
+
+		ENV_BUG_ON(ocf_alock_is_index_locked(alock, req, i));
+
+		if (!ocf_alock_lock_one_wr(alock, entry, cmpl, req, i)) {
+			ENV_BUG();
+			/* lock not acquired and not added to wait list */
+			ret = -OCF_ERR_NO_MEM;
+			goto err;
+		}
+	}
+
+	return ret;
+
+err:
+	for (; i >= 0; i--) {
+		entry = ocf_pio_lock_get_entry(alock, req, i);
+		if (entry == OUT_OF_RANGE)
+			continue;
+
+		ocf_alock_waitlist_remove_entry(alock, req, i, entry, OCF_WRITE);
+	}
+
+	return ret;
+}
+
+static struct ocf_alock_lock_cbs ocf_pio_conc_cbs = {
+		.lock_entries_fast = ocf_pio_lock_fast,
+		.lock_entries_slow = ocf_pio_lock_slow
+};
+
+int ocf_pio_async_lock(struct ocf_alock *alock, struct ocf_request *req,
+		ocf_req_async_lock_cb cmpl)
+{
+	return ocf_alock_lock_wr(alock, req, cmpl);
+}
+
+void ocf_pio_async_unlock(struct ocf_alock *alock, struct ocf_request *req)
+{
+	ocf_cache_line_t entry;
+	int i;
+
+	for (i = 0; i < req->core_line_count; i++) {
+		if (!ocf_alock_is_index_locked(alock, req, i))
+			continue;
+
+		entry = ocf_pio_lock_get_entry(alock, req, i);
+		if (entry == OUT_OF_RANGE)
+			continue;
+
+		ocf_alock_unlock_one_wr(alock, entry);
+		ocf_alock_mark_index_locked(alock, req, i, false);
+	}
+
+	req->alock_status = 0;
+}
+
+#define ALLOCATOR_NAME_FMT "ocf_%s_pio_conc"
+#define ALLOCATOR_NAME_MAX (sizeof(ALLOCATOR_NAME_FMT) + OCF_CACHE_NAME_SIZE)
+
+int ocf_pio_concurrency_init(struct ocf_alock **self, ocf_cache_t cache)
+{
+	struct ocf_metadata_ctrl *ctrl = cache->metadata.priv;
+	struct ocf_alock *alock;
+	struct pio_ctx *pio_ctx;
+	size_t base_size = ocf_alock_obj_size();
+	char name[ALLOCATOR_NAME_MAX];
+	int ret;
+	uint32_t pages_to_alloc = 0;
+	enum ocf_metadata_segment_id update_segments[] = {
+		metadata_segment_sb_config,
+		metadata_segment_part_config,
+		metadata_segment_core_config,
+		metadata_segment_core_uuid,
+		metadata_segment_collision,
+	};
+	int i;
+
```

Comment:
maybe add BUG_ON(ARRAY_SIZE(update_segments) > ARRAY_SIZE(pio_ctx->segments))

Reply:
I added build bugon

Reply:
I'd also add "FIXME: pio_concurrency should be parametrizable with segments list", to not leave this BUG_ON() as a lone monument of poor design.